Export PPL Command docs from SQL plugin repo (#11688)
opensearch-project/documentation-website: _sql-and-ppl/ppl/commands/stats.md

---
layout: default
title: stats
parent: Commands
grand_parent: PPL
nav_order: 38
---

# stats

The `stats` command calculates aggregations on the search results.

> ## Comparing `stats`, `eventstats`, and `streamstats`

Reviewer: ```suggestion
## Comparing stats, eventstats, and streamstats
```

For a comprehensive comparison of `stats`, `eventstats`, and `streamstats` commands, including their differences in transformation behavior, output format, aggregation scope, and use cases, see [Comparing stats, eventstats, and streamstats]({{site.url}}{{site.baseurl}}/sql-and-ppl/ppl/commands/streamstats/#comparing-stats-eventstats-and-streamstats).

## Syntax

The `stats` command has the following syntax:

```sql
stats [bucket_nullable=bool] <aggregation>... [by-clause]
```

## Parameters

The `stats` command supports the following parameters.

| Parameter | Required/Optional | Description |
| --- | --- | --- |
| `<aggregation>` | Required | An aggregation function. |
| `<by-clause>` | Optional | Groups results by specified fields or expressions. Syntax: `by [span-expression,] [field,]...` If no `by-clause` is specified, the stats command returns only one row, which is the aggregation over the entire search results. |

> | `bucket_nullable` | Optional | Controls whether to include `null` buckets in group-by aggregations. When `false`, ignores records in which the group-by field is null, resulting in faster performance. Default is the value of `plugins.ppl.syntax.legacy.preferred`. |

Reviewer: ```suggestion
| `bucket_nullable` | Optional | Controls whether to include `null` buckets in group-by aggregations. When `false`, ignores records in which the `group-by` field is null, resulting in faster performance. Default is the value of `plugins.ppl.syntax.legacy.preferred`. |
```

  

## Example 4: Calculate the average, sum, and count of a field by group  

The following query calculates the average age, sum of ages, and count of events for all accounts, grouped by gender:
  
```sql
source=accounts
| stats avg(age), sum(age), count() by gender
```
{% include copy.html %}
  
The query returns the following results:
  
| avg(age) | sum(age) | count() | gender |
| --- | --- | --- | --- |
| 28.0 | 28 | 1 | F |
| 33.666666666666664 | 101 | 3 | M |
  

## Example 5: Calculate the maximum of a field  

The following query calculates the maximum age for all accounts:
  
```sql
source=accounts
| stats max(age)
```
{% include copy.html %}
  
The query returns the following results:
  
| max(age) |
| --- |
| 36 |
  

## Example 6: Calculate the maximum and minimum of a field by group  

The following query calculates the maximum and minimum ages for all accounts, grouped by gender:
  
```sql
source=accounts
| stats max(age), min(age) by gender
```
{% include copy.html %}
  
The query returns the following results:
  
| max(age) | min(age) | gender |
| --- | --- | --- |
| 28 | 28 | F |
| 36 | 32 | M |
  

## Example 7: Calculate the distinct count of a field  

> To retrieve the count of distinct values of a field, you can use `DISTINCT_COUNT` (or `DC`) function instead of `COUNT`. The following query calculates both the count and the distinct count of the `gender` field for all accounts:

Reviewer: ```suggestion
To retrieve the count of distinct values of a field, you can use the `DISTINCT_COUNT` (or `DC`) function instead of `COUNT`. The following query calculates both the count and the distinct count of the `gender` field for all accounts:
```

The following query retrieves the count of `age` values grouped into 10-year intervals:
  
```sql
source=accounts
| stats count(age) by span(age, 10) as age_span
```
{% include copy.html %}
  
The query returns the following results:
  
| count(age) | age_span |
| --- | --- |
| 1 | 20 |
| 3 | 30 |
  

## Example 9: Calculate the count by a gender and span  

The following query retrieves the count of `age` grouped into 5-year intervals and broken down by `gender`:
  
```sql
source=accounts
| stats count() as cnt by span(age, 5) as age_span, gender
```
{% include copy.html %}
  
The query returns the following results:
  
| cnt | age_span | gender |
| --- | --- | --- |
| 1 | 25 | F |
| 2 | 30 | M |
| 1 | 35 | M |
  
The `span` expression is always treated as the first grouping key, regardless of its position in the `by` clause:
  
```sql
source=accounts
| stats count() as cnt by gender, span(age, 5) as age_span
```
{% include copy.html %}
  
The query returns the following results:
  
| cnt | age_span | gender |
| --- | --- | --- |
| 1 | 25 | F |
| 2 | 30 | M |
| 1 | 35 | M |
  

## Example 10: Count and retrieve email list by gender and age span

> The following query calculates the count of `age` values grouped into 5-year intervals and by `gender`, and also returns a list of up to 5 emails for each group:

Reviewer: ```suggestion
The following query calculates the count of `age` values grouped into 5-year intervals as well as by `gender` and also returns a list of up to 5 emails for each group:
```

```sql
source=accounts
| stats count() as cnt, take(email, 5) by span(age, 5) as age_span, gender
```
{% include copy.html %}
  
The query returns the following results:
  
| cnt | take(email, 5) | age_span | gender |
| --- | --- | --- | --- |
| 1 | [] | 25 | F |
| 2 | [amberduke@pyrami.com,daleadams@boink.com] | 30 | M |
| 1 | [hattiebond@netagy.com] | 35 | M |
  

## Example 11: Calculate the percentile of a field  

The following query calculates the 90th percentile of `age` for all accounts:
  
```sql
source=accounts
| stats percentile(age, 90)
```
{% include copy.html %}
  
The query returns the following results:
  
| percentile(age, 90) |
| --- |
| 36 |
  

## Example 12: Calculate the percentile of a field by group  

The following query calculates the 90th percentile of `age` for all accounts, grouped by `gender`:
  
```sql
source=accounts
| stats percentile(age, 90) by gender
```
{% include copy.html %}
  
The query returns the following results:
  
| percentile(age, 90) | gender |
| --- | --- |
| 28 | F |
| 36 | M |
  

## Example 13: Calculate the percentile by a gender and span  

> The following query calculates the 90th percentile of `age`, grouped into 10-year intervals and by `gender`:

Reviewer: ```suggestion
The following query calculates the 90th percentile of `age`, grouped into 10-year intervals as well as by `gender`:
```

### Bucket aggregation results may be approximate for high-cardinality fields

In OpenSearch, `doc_count` values for a `terms` bucket aggregation can be approximate. Thus, any aggregations (such as `sum` or `avg`) performed on those buckets may also be approximate.

For example, the following query retrieves the top 10 URLs:

```sql
source=hits
| stats bucket_nullable=false count() as c by URL
| sort - c
| head 10
```

{% include copy.html %}

This query is translated into a `terms` aggregation in OpenSearch with `"order": { "_count": "desc" }`. For fields with high cardinality, some buckets may be discarded, so the results may only be approximate.

### Sorting by `doc_count` in ascending order may produce inaccurate results

When retrieving the least frequent terms for high-cardinality fields, results may be inaccurate. Shard-level aggregations can miss globally rare terms or misrepresent their frequency, causing errors in the overall results.

For example, the following query retrieves the 10 least frequent URLs:

```sql
source=hits
| stats bucket_nullable=false count() as c by URL
| sort + c
| head 10
```

{% include copy.html %}

> A globally rare term might not appear as rare on every shard or could be entirely absent from some shard results. Conversely, a term infrequent on one shard might be common on another. In both cases, shard-level approximations can cause rare terms to be missed, leading to inaccurate overall results.

Reviewer: ```suggestion
A globally rare term might not appear as rare on every shard or could be entirely absent from some shard results. Conversely, a term that is infrequent on one shard might be common on another. In both cases, shard-level approximations can cause rare terms to be missed, leading to inaccurate overall results.
```
